File: 2015/lua/8/main.lua
local function countChars(s, counts)
    counts.codeChars = counts.codeChars + #s

    local i = 2 -- Skip first quote
    while i < #s do
        if s:sub(i, i) == '\\' then
            if i + 1 < #s then
                local nextChar = s:sub(i + 1, i + 1)
                if nextChar == '\\' or nextChar == '"' then
                    counts.stringChars = counts.stringChars + 1
                    i = i + 1
                elseif nextChar == 'x' then
                    if i + 3 < #s then
                        counts.stringChars = counts.stringChars + 1
                        i = i + 3
                    end
                else
                    counts.stringChars = counts.stringChars + 1
                end
            end
        else
            counts.stringChars = counts.stringChars + 1
        end
        i = i + 1
    end

    return counts
end

local function encodeString(s)
    local encoded = '"'
    for i = 1, #s do
        local ch = s:sub(i, i)
        if ch == '"' or ch == '\\' then
            encoded = encoded .. '\\'
        end
        encoded = encoded .. ch
    end
    encoded = encoded .. '"'
    return { encoded = encoded, length = #encoded }
end

local function main()
    -- get command line arguments
    local args = arg
    if #args < 1 then
        print("Usage: lua main.lua <input.txt>")
        return
    end

    local filename = args[1]

    local file = io.open(filename, "r")
    if not file then
        print("File not found: " .. filename)
        return
    end

    local data = file:read("*all")
    file:close()

    if not data then
        print("File is empty: " .. filename)
        return
    end

    local lines = {}
    for line in data:gmatch("[^\r\n]+") do
        table.insert(lines, line)
    end

    local counts = { codeChars = 0, stringChars = 0 }
    local encodedLength = 0

    for _, line in ipairs(lines) do
        counts = countChars(line, counts)
        local encoded = encodeString(line)
        encodedLength = encodedLength + encoded.length
    end

    print("Part 1:", counts.codeChars - counts.stringChars)
    print("Part 2:", encodedLength - counts.codeChars)
end

main()

File: 2015/javascript/8/main.js
const fs = require('fs');

function countChars(s, counts) {
    counts.codeChars += s.length;

    for (let i = 1; i < s.length - 1; i++) { // Skip first and last quote
        if (s[i] === '\\') {
            if (i + 1 < s.length - 1) {
                switch (s[i + 1]) {
                    case '\\':
                    case '"':
                        counts.stringChars++;
                        i++;
                        break;
                    case 'x':
                        if (i + 3 < s.length - 1) {
                            counts.stringChars++;
                            i += 3;
                        }
                        break;
                    default:
                        counts.stringChars++;
                }
            }
        } else {
            counts.stringChars++;
        }
    }

    return counts;
}

function encodeString(s) {
    let encoded = '"';
    for (let ch of s) {
        if (ch === '"' || ch === '\\') {
            encoded += '\\';
        }
        encoded += ch;
    }
    encoded += '"';
    return { encoded, length: encoded.length };
}

function main(){
    // get command line arguments
    const args = process.argv.slice(1);
    if (args.length < 2){
        console.log("Usage: node main.js <input.txt>");
        return;
    }

    const filename = args[1];

    if (!fs.existsSync(filename)){
        console.log("File not found: " + filename);
        return;
    }

    const data = fs.readFileSync(filename, 'utf8');

    if (!data){
        console.log("File is empty: " + filename);
        return;
    }

    const lines = data.split('\n').filter(line => line.trim().length > 0);

    let counts = { codeChars: 0, stringChars: 0 };
    let encodedLength = 0;
    
    for (let line of lines) {
        counts = countChars(line, counts);
        const { encoded, length } = encodeString(line);
        encodedLength += length;
    }

    console.log("Part 1:", counts.codeChars - counts.stringChars);

    console.log("Part 2:", encodedLength - counts.codeChars);
}


main();